File: frontend/src/lib/paintedEdges.js
// Candidate detection is NOT face detection: text never enables a texture.
const FEATURE = /\b(?:(?:cantos?|bordes?)\s+(?:de\s+(?:las?\s+)?paginas?\s+)?(?:pintad[oa]s?|tintad[oa]s?|decorad[oa]s?|ilustrad[oa]s?|metalizad[oa]s?|coloread[oa]s?|tenid[oa]s?|dorad[oa]s?|platead[oa]s?|estampad[oa]s?)|taco\s+pintad[oa]s?|(?:sprayed|stencilled|stenciled|painted|decorated|gilded|stained|coloured|colored)\s+(?:book\s+)?edges)\b/g;
const NEGATION = /(?:\bsin(?:\s+ningun[oa]?)?|\bno\s+(?:tiene|incluye|lleva|trae|presenta|son|estan)(?:\s+\w+){0,3})\s*$/;
const SPECIAL = /\b(?:edicion\s+(?:especial|limitada|coleccionista)|coleccionista|collector(?:'s)?\s+edition|deluxe\s+edition)\b/;
const CONDITIONAL = /\b(?:primera\s+(?:edicion|tirada|impresion)|hasta\s+(?:agotar|fin\s+de)\s+(?:existencias|stock)|no\s+(?:se\s+)?(?:puede\s+)?garantiza\w*|solo\s+(?:en\s+)?(?:la\s+)?primera)\b/;
const PROVISIONAL = /\b(?:provisional|diseno\s+no\s+definitivo|artwork\s+(?:not\s+final|subject\s+to\s+change))\b/;

function text(value) {
  return String(value || "").replace(/<[^>]*>/g, " ").normalize("NFD").replace(/[\u0300-\u036f]/g, "").toLowerCase().replace(/\s+/g, " ").trim();
}

function validIsbn(isbn) {
  if (/^\d{13}$/.test(isbn)) {
    const total = [...isbn.slice(0, 12)].reduce((sum, n, i) => sum + Number(n) * (i % 2 ? 3 : 1), 0);
    return (10 - total % 10) % 10 === Number(isbn[12]);
  }
  if (/^\d{9}[\dX]$/.test(isbn)) return [...isbn].reduce((sum, n, i) => sum + (n === "X" ? 10 : Number(n)) * (10 - i), 0) % 11 === 0;
  return false;
}

export function detectPaintedEdges(item = {}) {
  const title = text([item.title || item.titulo, item.edition_label || item.edition || item.edicion, item.special_binding || item.specialBinding || item.encuadernacion_especial].filter(Boolean).join(" "));
  const synopsis = text(item.synopsis || item.sinopsis);
  const fields = [title, synopsis];
  const all = fields.join(" ");
  const matches = fields.flatMap((field, index) => [...field.matchAll(FEATURE)]
    .filter((match) => !NEGATION.test(field.slice(Math.max(0, match.index - 60), match.index)))
    .map((match) => ({ evidence: match[0], title: index === 0 })));
  const candidate = matches.length > 0 || SPECIAL.test(title);
  if (!candidate) return { status: "none", reason: "no_signal", evidence: "" };
  if (/\b(?:estuche|box\s*set|boxed\s*set|cofre)\b/.test(title)) return { status: "excluded", reason: "box_set", evidence: matches[0]?.evidence || "" };
  const evidence = matches[0]?.evidence || "";
  const review = (reason) => ({ status: "review", reason, evidence });
  const isbn = String(item.isbn || "").toUpperCase().replace(/[^0-9X]/g, "");
  if (!validIsbn(isbn)) return review("isbn_required");
  const cover = String(item.cover || item.imagen_portada || "");
  const photoIsbn = cover.match(/casadellibro\.com\/[^?#]*\/(\d{13})\.(?:webp|jpe?g|png)(?:[?#]|$)/i)?.[1];
  if (photoIsbn && photoIsbn !== isbn) return review("cover_mismatch");
  if (PROVISIONAL.test(all)) return review("provisional");
  if (matches.length && CONDITIONAL.test(all)) return review("first_printing");
  if (matches.some((match) => match.title)) return { status: "explicit", reason: "edition_label", evidence };
  if (matches.length) return review("synopsis_only");
  return { status: "possible", reason: "special_edition", evidence: "" };
}

export function paintedEdgeReviewMessage(result) {
  const messages = {
    edition_label: "Cantos indicados en esta edición. Revisa la foto y delimita el canto visible antes de aplicarlo.",
    synopsis_only: "La sinopsis menciona cantos especiales: confirma que pertenecen a este ISBN y no a otra edición.",
    first_printing: "Los cantos pueden limitarse a la primera tirada o al stock disponible. Comprueba esta edición.",
    provisional: "El dibujo del canto es provisional. Espera una foto definitiva antes de aplicarlo.",
    isbn_required: "Falta un ISBN válido para identificar la edición con cantos especiales.",
    cover_mismatch: "La imagen de portada corresponde a otro ISBN. Corrige la foto antes de preparar el canto.",
    special_edition: "Edición especial por comprobar: no implica que tenga cantos pintados.",
    box_set: "Estuche con varios libros: revisa cada volumen por separado; no se aplicará un único canto al conjunto.",
  };
  return messages[result?.reason] || "";
}

export function discardStaleFaceSelection(item, changes) {
  const identityChanged = ["isbn", "cover"].some((field) => Object.hasOwn(changes, field) && changes[field] !== item[field]);
  return identityChanged ? { ...changes, visual: null } : changes;
}

File: tools/scraping_libros/libros/painted_edges.py
"""Conservative edition candidates. Text never supplies face coordinates."""
import re
import unicodedata

FEATURE = re.compile(
    r"\b(?:(?:cantos?|bordes?)\s+(?:de\s+(?:las?\s+)?paginas?\s+)?"
    r"(?:pintad[oa]s?|tintad[oa]s?|decorad[oa]s?|ilustrad[oa]s?|metalizad[oa]s?|"
    r"coloread[oa]s?|tenid[oa]s?|dorad[oa]s?|platead[oa]s?|estampad[oa]s?)|"
    r"taco\s+pintad[oa]s?|(?:sprayed|stencilled|stenciled|painted|decorated|gilded|"
    r"stained|coloured|colored)\s+(?:book\s+)?edges)\b"
)
NEGATION = re.compile(r"(?:\bsin(?:\s+ningun[oa]?)?|\bno\s+(?:tiene|incluye|lleva|trae|presenta|son|estan)(?:\s+\w+){0,3})\s*$")
SPECIAL = re.compile(r"\b(?:edicion\s+(?:especial|limitada|coleccionista)|coleccionista|collector(?:'s)?\s+edition|deluxe\s+edition)\b")
CONDITIONAL = re.compile(r"\b(?:primera\s+(?:edicion|tirada|impresion)|hasta\s+(?:agotar|fin\s+de)\s+(?:existencias|stock)|no\s+(?:se\s+)?(?:puede\s+)?garantiza\w*|solo\s+(?:en\s+)?(?:la\s+)?primera)\b")
PROVISIONAL = re.compile(r"\b(?:provisional|diseno\s+no\s+definitivo|artwork\s+(?:not\s+final|subject\s+to\s+change))\b")


def text(value):
    value = re.sub(r"<[^>]*>", " ", str(value or ""))
    value = "".join(c for c in unicodedata.normalize("NFD", value) if not unicodedata.combining(c))
    return re.sub(r"\s+", " ", value.lower()).strip()


def valid_isbn(value):
    if re.fullmatch(r"\d{13}", value):
        total = sum(int(n) * (3 if i % 2 else 1) for i, n in enumerate(value[:12]))
        return (10 - total % 10) % 10 == int(value[12])
    if re.fullmatch(r"\d{9}[\dX]", value):
        return sum((10 if n == "X" else int(n)) * (10 - i) for i, n in enumerate(value)) % 11 == 0
    return False


def detect_painted_edges(item):
    title = text(" ".join(str(v) for v in [item.get("title") or item.get("titulo"), item.get("edition_label") or item.get("edition") or item.get("edicion")] if v))
    synopsis = text(item.get("synopsis") or item.get("sinopsis"))
    fields = [title, synopsis]
    all_text = " ".join(fields)
    matches = [(match.group(), index == 0) for index, field in enumerate(fields)
               for match in FEATURE.finditer(field)
               if not NEGATION.search(field[max(0, match.start() - 60):match.start()])]
    if not matches and not SPECIAL.search(title):
        return {"status": "none", "reason": "no_signal", "evidence": ""}
    evidence = matches[0][0] if matches else ""
    if re.search(r"\b(?:estuche|box\s*set|boxed\s*set|cofre)\b", title):
        return {"status": "excluded", "reason": "box_set", "evidence": evidence}
    def review(reason):
        return {"status": "review", "reason": reason, "evidence": evidence}
    isbn = re.sub(r"[^0-9X]", "", str(item.get("isbn") or "").upper())
    if not valid_isbn(isbn):
        return review("isbn_required")
    cover = str(item.get("cover") or item.get("imagen_portada") or "")
    photo_isbn = re.search(r"casadellibro\.com\/[^?#]*\/(\d{13})\.(?:webp|jpe?g|png)(?:[?#]|$)", cover, re.I)
    if photo_isbn and photo_isbn.group(1) != isbn:
        return review("cover_mismatch")
    if PROVISIONAL.search(all_text):
        return review("provisional")
    if matches and CONDITIONAL.search(all_text):
        return review("first_printing")
    if any(is_title for _, is_title in matches):
        return {"status": "explicit", "reason": "edition_label", "evidence": evidence}
    if matches:
        return review("synopsis_only")
    return {"status": "possible", "reason": "special_edition", "evidence": ""}
